### PR TITLE
Add `array.reverse(arr)` and `strings.reverse(string)` built-ins

### LINKED ISSUE #3736: Add a built-in function to reverse arrays and strings
OPA does not currently include a built-in function for reversing strings and arrays. While it's possible to write a comprehension that does this, it's a bit cumbersome. We can add two simple built-in functions for reversing strings and arrays:

```
strings.reverse(str) => str
array.reverse(arr) => arr
```

### PR DESCRIPTION
**Pull request by kristiansvalland**
This PR adds the `array.reverse(arr)` and `strings.reverse(string)` built-ins.

Specifically, the reverse functionality
* reverses arrays element-wise, and
* reverses strings character-wise (rune in go terms)

Fixes #3736

### CODE REVIEW

**Comment by kristiansvalland** [comment]
I would like input on the steps needed to make the "Wasm Support" claim I made in the docs true. I feel that this should be easily solvable since no complex dependencies are used in the implementation, but I do not know how to. Is there a recipe for auto-generating the WASM code, or must I write this manually?

File: ast/builtins.go [hunk 1]
```diff
@@ -1080,6 +1093,17 @@ var Sprintf = &Builtin{
 	),
 }
 
+// Reverse returns the given string, reversed.
+var Reverse = &Builtin{
+	Name: "reverse",
```
Comment (anderseknert) [inline ast/builtins.go:1098]: I wonder if we should go with `strings.reverse` instead? There's currently only one string function using that prefix, but I think we'd like to avoid using common names as built-ins, plus the `strings.` part would make it obvious what was expected. We currently only do this for `strings.replace_n` though. WDYT @srenatus ?

File: docs/content/policy-reference.md [hunk 2]
```diff
@@ -324,6 +324,7 @@ complex types.
 | Built-in | Description | Wasm Support |
 | ------- |-------------|---------------|
 | <span class="opa-keep-it-together">``output := array.concat(array, array)``</span> | ``output`` is the result of concatenating the two input arrays together. | ✅ |
+| <span class="opa-keep-it-together">``output := array.reverse(array)``</span> | ``output`` is the result of reversing the order of the elements in ``array``. | ✅ |
```
Comment (anderseknert) [inline docs/content/policy-reference.md:327]: This doesn't have Wasm support yet, so should be "SDK-dependent"

File: docs/content/policy-reference.md [hunk 3]
```diff
@@ -379,6 +380,7 @@ complex types.
 | <span class="opa-keep-it-together">``output := indexof(string, search)``</span> | ``output`` is the index inside ``string`` where ``search`` first occurs, or -1 if ``search`` does not exist | ✅ |
 | <span class="opa-keep-it-together">``output := lower(string)``</span> | ``output`` is ``string`` after converting to lower case | ✅ |
 | <span class="opa-keep-it-together">``output := replace(string, old, new)``</span> | ``output`` is a ``string`` representing ``string`` with all instances of ``old`` replaced by ``new`` | ✅ |
+| <span class="opa-keep-it-together">``output := reverse(string)``</span> | ``output`` is ``string`` reversed | ✅ |
```
Comment (anderseknert) [inline docs/content/policy-reference.md:383]: Same

File: test/cases/testdata/array/test-array-0052.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    foo:
+    - 1
+    - 2
+    - 3
+  modules:
+  - |
+    package generated
+
+    p = x {
+      array.reverse(data.foo, __local0__)
```
Comment (anderseknert) [inline test/cases/testdata/array/test-array-0052.yaml:14]: Do we need `__local0__` here? 

Would not
```rego
p = x {
      x := array.reverse(data.foo)
}
```
or even

```rego
p := array.reverse(data.foo)
```
suffice?

**Comment by srenatus** [comment]
https://github.com/open-policy-agent/opa/issues/4140 all that's written here should apply in this case, too. Let me know if you get stuck.

File: test/cases/testdata/array/test-array-0052.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+cases:
+- note: 'array/reverse_123'
+  query: data.generated.p = x
```
Comment (anderseknert) [inline test/cases/testdata/array/test-array-0052.yaml:3]: I'd name this something else than `generated`

File: test/cases/testdata/strings/test-strings-0924.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
+cases:
```
Comment (anderseknert) [inline test/cases/testdata/strings/test-strings-0924.yaml:1]: These tests are good, but it would be good to see a case or two with unicode characters. Should work given your implementation :)

**Review by anderseknert — COMMENTED** [review]
Looks good! Left a couple of comments.. and we're gonna want to squash those commits into a single, even though I guess you branched off of @olamiko's previous work.

File: ast/builtins.go (hunk 1 above)
Comment (kristiansvalland) [reply ast/builtins.go]: I like the idea, so I went ahead and changed it to what you suggested.
A `reverse(something)` function in a policy would probably require some unnecessary thinking for the readers of said policy, when `strings.reverse` is obvious what it does and what type the argument is/should be.
Also a `reverse` function could maybe lead people to ask "What else does it reverse?", leading to frustration when it only works on strings.

File: docs/content/policy-reference.md (hunk 2 above)
Comment (kristiansvalland) [reply docs/content/policy-reference.md:327]: I will see if I can make this claim true, but I will of course change this if it is does not get included in this PR.

File: test/cases/testdata/array/test-array-0052.yaml (hunk 4 above)
Comment (kristiansvalland) [reply test/cases/testdata/array/test-array-0052.yaml]: Totally. Changed to that now. Thanks 😄

File: test/cases/testdata/array/test-array-0052.yaml (hunk 5 above)
Comment (kristiansvalland) [reply test/cases/testdata/array/test-array-0052.yaml]: Changed to `test`

File: test/cases/testdata/strings/test-strings-0924.yaml (hunk 6 above)
Comment (kristiansvalland) [reply test/cases/testdata/strings/test-strings-0924.yaml:1]: I added two unicode test cases. One that is quite straight forward I would say, and another that might have unexpected results (strings/reverse_unicode_multi_char_emojii). This is when a unicode glyph(?) consists of multiple codepoints, which is often the case for modern emojies. As the functionality is now, we reverse at the level of codepoints, which changes how the emojies are rendered.

Is the current functionality as expected? Or would we want to reorder at the glyph level instead?

Comment (anderseknert) [reply test/cases/testdata/strings/test-strings-0924.yaml:1]: Some good research there! I'd say reordering at glyph level would be less surprising, or what do you think? It would obviously be (happily!) surprising if we saw policies relying on the reverse implementation of emojis 😆 but is this not case with e.g. characters like the German ß as well?

File: topdown/array.go [hunk 7]
```diff
@@ -71,7 +71,25 @@ func builtinArraySlice(a, i, j ast.Value) (ast.Value, error) {
...
+
+	length := arr.Len()
+	reversedArr := make([]*ast.Term, length)
+
+	for index := 0; index < length; index++ {
+		reversedArr[index] = arr.Elem(length - index - 1)
+	}
+
+	return iter(ast.ArrayTerm(reversedArr...))
+
```
Comment (anderseknert) [inline topdown/array.go:88]: Can we skip the extra newline after the return here and in `builtinReverse` below?

**Review by anderseknert — COMMENTED** [review]
Good changes. We're getting there!

File: test/cases/testdata/strings/test-strings-0924.yaml (hunk 6 above)
Comment (kristiansvalland) [reply test/cases/testdata/strings/test-strings-0924.yaml:1]: I do agree that that would be the least surprising. I think the german ß is a single unicode code point (U+00DF), but some characters with accents etc. are examples of characters that are modified by following code points (there are some examples in the repository linked to below). 

After doing some research on the matter yesterday, I do believe that that this is more complicated than what I would have guessed beforehand. I'd like to present some findings on the matter to spark some discussion:

- A visual character as interpreted by a human is called a grapheme cluster (not glyph as I first suggested) in unicode therms. What we want to accomplish is to reverse at the level of grapheme. To cite from https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries:
  > It is important to recognize that what the user thinks of as a “character" [...] this is called a user-perceived character. [...] These user-perceived characters are approximated by what is called a grapheme cluster, which can be determined programmatically
- I have not yet read said document in full, but I believe it details how grapheme-cluster boundaries can be determined. I've also found an example of a public go-library which does this: https://github.com/rivo/uniseg. I am not suggesting that this is included as a dependency, but rather want to provide it as an example.
- Implementing the strings.reverse this way would (arguably) be inconsistent with current implementations of string operations, like for instance split. Take a look at this example policy: https://play.openpolicyagent.org/p/aYjrhAiKJu. Is this also non-surprising?
- If we implement some way of doing grapheme-clustering in the strings.reverse function, both in go and wasm, would this consequently lead to continuous maintanence need of said functionality as unicode is expanded with new emojies and characters in the future? 

To summarise based on what I've written above: Unicode handling is hard, and in some cases depending on the use-case. 
What would be the best way forward on this matter?

File: docs/content/policy-reference.md (hunk 2 above)
Comment (kristiansvalland) [reply docs/content/policy-reference.md:327]: Based on @srenatus pointer above, I've tried adding implementations of WASM for both functions with some tests. Let me know if you see some problems or more tests are needed.

Comment (anderseknert) [reply docs/content/policy-reference.md:327]: Awesome! @srenatus will be better equipped to review that part :)

File: test/cases/testdata/strings/test-strings-0924.yaml (hunk 6 above)
Comment (anderseknert) [reply test/cases/testdata/strings/test-strings-0924.yaml:1]: Oh wow, yeah the flags look like they're always composed of multiple code points. Funny thing that when you try out a policy with those included in the Playground.. on Firefox you can actually see the cursor moving halfway inside of the flag when pressing the right arrow, and you'll need to press twice to make it to the next one 😄 

Looks like we've pretty much just been lucky in previous unicode and character position tests like this one https://github.com/open-policy-agent/opa/blob/main/test/cases/testdata/strings/test-strings-indexof-unicode.yaml

Let's ignore these for now. As long as any "normal" characters (i.e. those found in common languages and likely to be prevalent in policy) works as expected, I think that's good enough. I'm sure there will be a trendy new startup somewhere creating policies exlusively using emojis, but if that's so I don't think it would be unfair to ask them for help with supporting their use case :)

**Review by anderseknert — APPROVED** [review]
LGTM! Awaiting @srenatus to review the Wasm bits.

Also, if you could squash the commits now, that would be helpful.

File: test/cases/testdata/strings/test-strings-0924.yaml (hunk 6 above)
Comment (kristiansvalland) [reply test/cases/testdata/strings/test-strings-0924.yaml:1]: That's interesting, and even shows that established software like Firefox can have "surprises" when it comes to unicode handling.

Yep, for instance if one of the preceding emojis in "strings/indexof: unicode string emoji" were substituted with a flag, the want_result would have to be incremented by 1.

Hehe, that's a fair point.
Just to summarise what we can expect: Ascii should work without any surprises, and most latin-based characters I think, though there will be some exceptions like "ä" vs "ä". They look identical (on my computer at least), but the first is one code point, whereas the latter is two. Here is another example from the playground (using split, but the inner workings is the same as for strings.reverse): https://play.openpolicyagent.org/p/rvzH7nbNQW 
If "äe" (latter) would be reversed, I guess it will become "ëa", but if "äe" (first) is reversed, it becomes "eä". But as you say, characters such as these are probably not expected to be prevalent in policies.

Comment (anderseknert) [reply test/cases/testdata/strings/test-strings-0924.yaml:1]: Agreed, let's make a best effort (which you've definitely done already!) and improve on that later if requested.

**Review by anderseknert — APPROVED** [review]
LGTM! Thanks for this contribution, and for helping us understand the nuances of unicode :)

**Comment by kristiansvalland** [comment]
Thanks for you review and feedback, Anders!

**Review by srenatus — APPROVED** [review]
👏 thanks a lot 🎉
